File: app/src/main/java/com/minerdev/greformanager/HouseListAdapter.java
package com.minerdev.greformanager;

import android.net.Uri;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.DiffUtil;
import androidx.recyclerview.widget.ListAdapter;
import androidx.recyclerview.widget.RecyclerView;

import com.bumptech.glide.Glide;

import java.sql.Date;

public class HouseListAdapter extends ListAdapter<House, HouseListAdapter.ViewHolder> {
    private OnItemClickListener listener;

    public HouseListAdapter(DiffUtil.ItemCallback<House> diffCallback) {
        super(diffCallback);
    }

    @NonNull
    @Override
    public ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        LayoutInflater inflater = LayoutInflater.from(parent.getContext());
        View itemView = inflater.inflate(R.layout.item_house, parent, false);
        return new ViewHolder(itemView, listener);
    }

    @Override
    public void onBindViewHolder(@NonNull ViewHolder holder, int position) {
        House item = getItem(position);
        holder.setItem(item);
    }

    public House get(int position) {
        return getItem(position);
    }

    public void setOnItemClickListener(OnItemClickListener clickListener) {
        listener = clickListener;
    }

    public interface OnItemClickListener {
        void onItemClick(HouseListAdapter.ViewHolder viewHolder, View view, int position);
    }

    public static class ViewHolder extends RecyclerView.ViewHolder {
        private final TextView textView_payment;
        private final TextView textView_price;
        private final TextView textView_house_info;
        private final TextView textView_description;
        private final TextView textView_upload_time;
        private final ImageView imageView_profile;

        public ViewHolder(@NonNull View itemView, final OnItemClickListener clickListener) {
            super(itemView);

            textView_payment = itemView.findViewById(R.id.houseItem_textView_payment_type);
            textView_price = itemView.findViewById(R.id.houseItem_textView_price);
            textView_house_info = itemView.findViewById(R.id.houseItem_textView_house_info);
            textView_description = itemView.findViewById(R.id.houseItem_textView_description);
            textView_upload_time = itemView.findViewById(R.id.houseItem_textView_upload_time);
            imageView_profile = itemView.findViewById(R.id.houseItem_imageView_profile);

            LinearLayout linearLayout = itemView.findViewById(R.id.houseItem_layout);
            linearLayout.setOnClickListener(v -> {
                if (clickListener != null) {
                    clickListener.onItemClick(ViewHolder.this, itemView, getAdapterPosition());
                }
            });
        }

        public void setItem(House house) {
            HouseWrapper houseWrapper = new HouseWrapper(house);

            textView_payment.setText(houseWrapper.getPaymentType());
            textView_price.setText(houseWrapper.getPrice());
            textView_house_info.setText(houseWrapper.getHouseInfo());
            textView_description.setText(houseWrapper.getDetailInfo());
            Date date = new Date(house.created_at);
            textView_upload_time.setText(date.toString());

            Uri uri = Uri.parse(Constants.getInstance().DNS + "/storage/images/" + house.id + "/" + house.thumbnail);
            Glide.with(itemView).load(uri).into(imageView_profile);
        }
    }

    public static class DiffCallback extends DiffUtil.ItemCallback<House> {
        @Override
        public boolean areItemsTheSame(@NonNull House oldItem, @NonNull House newItem) {
            return oldItem == newItem;
        }

        @Override
        public boolean areContentsTheSame(@NonNull House oldItem, @NonNull House newItem) {
            return oldItem.id == newItem.id;
        }
    }
}

File: app/src/main/java/com/minerdev/greformanager/HouseListAdapter.kt
package com.minerdev.greformanager

import android.net.Uri
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.LinearLayout
import android.widget.TextView
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.ListAdapter
import androidx.recyclerview.widget.RecyclerView
import com.bumptech.glide.Glide

class HouseListAdapter(diffCallback: DiffCallback) : ListAdapter<House, HouseListAdapter.ViewHolder>(diffCallback) {
    private var listener: OnItemClickListener? = null

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val inflater = LayoutInflater.from(parent.context)
        val itemView = inflater.inflate(R.layout.item_house, parent, false)
        return ViewHolder(itemView, listener)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        val item = getItem(position)
        holder.bind(item!!)
    }

    operator fun get(position: Int): House? {
        return getItem(position)
    }

    fun setOnItemClickListener(clickListener: OnItemClickListener) {
        listener = clickListener
    }

    interface OnItemClickListener {
        fun onItemClick(viewHolder: ViewHolder, view: View, position: Int)
    }

    class ViewHolder(itemView: View, clickListener: OnItemClickListener?) : RecyclerView.ViewHolder(itemView) {
        private val tvPaymentType: TextView = itemView.findViewById(R.id.tv_payment_type)
        private val tvPrice: TextView = itemView.findViewById(R.id.tv_price)
        private val tvHouseInfo: TextView = itemView.findViewById(R.id.tv_house_info)
        private val tvDescription: TextView = itemView.findViewById(R.id.tv_description)
        private val tvUploadTime: TextView = itemView.findViewById(R.id.tv_upload_time)
        private val ivProfile: ImageView = itemView.findViewById(R.id.iv_profile)

        fun bind(house: House) {
            val houseWrapper = HouseWrapper(house)

            tvPaymentType.text = houseWrapper.paymentType
            tvPrice.text = houseWrapper.price
            tvHouseInfo.text = houseWrapper.houseInfo
            tvDescription.text = houseWrapper.detailInfo
            tvUploadTime.text = AppHelper.instance.getDiffTimeMsg(house.created_at)

            val uri = Uri.parse(Constants.instance.DNS + "/storage/images/" + house.id + "/" + house.thumbnail)
            Glide.with(itemView).load(uri).into(ivProfile)
        }

        init {
            val linearLayout = itemView.findViewById<LinearLayout>(R.id.houseItem_layout)
            linearLayout.setOnClickListener {
                clickListener?.onItemClick(this@ViewHolder, itemView, bindingAdapterPosition)
            }
        }
    }

    class DiffCallback : DiffUtil.ItemCallback<House>() {
        override fun areItemsTheSame(oldItem: House, newItem: House): Boolean {
            return oldItem.id == newItem.id
        }

        override fun areContentsTheSame(oldItem: House, newItem: House): Boolean {
            return oldItem == newItem
        }
    }
}